File: LC-36-valid_sudoku-arrays_hashing/valid_sudoku.py
'''
Determine if a 9 x 9 Sudoku board is valid. Only the filled cells need to be validated according to the following rules:

Each row must contain the digits 1-9 without repetition.
Each column must contain the digits 1-9 without repetition.
Each of the nine 3 x 3 sub-boxes of the grid must contain the digits 1-9 without repetition.

Note:
A Sudoku board (partially filled) could be valid but is not necessarily solvable.
Only the filled cells need to be validated according to the mentioned rules.
'''
class Solution:
    def isValidSudoku(self, board: list) -> bool:
        n = 9
        rows = [set() for _ in range(n)]
        cols = [set() for _ in range(n)]
        boxs = [set() for _ in range(n)]

        for row_idx in range(n):
            for col_idx in range(n):
                val = board[row_idx][col_idx]
                if val == ".":
                    continue
                #check row
                if val in rows[row_idx]:
                    return False
                rows[row_idx].add(val)
                #check column
                if val in cols[col_idx]:
                    return False
                cols[col_idx].add(val)
                #check box
                box_idx = (row_idx//3) * 3 + (col_idx//3)
                if val in boxs[box_idx]:
                    return False
                boxs[box_idx].add(val)
                
        return True

File: LC-36-valid_sudoku-arrays_hashing/valid_sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function (board) {
  let n = 9;
  let rows = [];
  let cols = [];
  let boxs = [];

  for (i = 0; i < n; i++) {
    let rowi = new Set();
    let coli = new Set();
    let boxi = new Set();
    rows.push(rowi);
    cols.push(coli);
    boxs.push(boxi);
  }

  for (row_i = 0; row_i < n; row_i++) {
    for (col_i = 0; col_i < n; col_i++) {
      let val = board[row_i][col_i];
      if (val == ".") {
        continue;
      }
      let box_i = Math.floor(row_i / 3) * 3 + Math.floor(col_i / 3);
      switch (true) {
        case rows[row_i].has(val):
          return false;

        case cols[col_i].has(val):
          return false;

        case boxs[box_i].has(val):
          return false;

        default:
          rows[row_i].add(val);
          cols[col_i].add(val);
          boxs[box_i].add(val);
      }
    }
  }
  return true;
};
